runtime/interpreter/deployedcontract.go

```diff
@@ -37,18 +41,62 @@ func NewDeployedContractValue(
 	name *StringValue,
 	code *ArrayValue,
 ) *SimpleCompositeValue {
+	publicTypesFuncValue := newPublicTypesFunctionValue(inter, address, name)
 	return NewSimpleCompositeValue(
 		inter,
 		sema.DeployedContractType.TypeID,
 		deployedContractStaticType,
 		deployedContractFieldNames,
 		map[string]Value{
-			sema.DeployedContractTypeAddressFieldName: address,
-			sema.DeployedContractTypeNameFieldName:    name,
-			sema.DeployedContractTypeCodeFieldName:    code,
+			sema.DeployedContractTypeAddressFieldName:        address,
+			sema.DeployedContractTypeNameFieldName:           name,
+			sema.DeployedContractTypeCodeFieldName:           code,
+			sema.DeployedContractTypePublicTypesFunctionName: publicTypesFuncValue,
 		},
 		nil,
 		nil,
 		nil,
 	)
 }
+
+func newPublicTypesFunctionValue(inter *Interpreter, addressValue AddressValue, name *StringValue) FunctionValue {
+	// public types only need to be computed once per contract
+	var once sync.Once
```

Comment:
No need to make this thread-safe. Maybe just replace by initializing when `publicTypes == nil`